### DISCUSSION TITLE
sporadical package-lock.json diff since using binarySource=install

### DISCUSSION THREAD

**Discussion by herndlm** [renovatebot/renovate#14562] [Request Help] [answered]
We recently upgraded some package-lock.json files from version 1 to version 2 via updating Node 14 to Node 16. While doing that I enabled `binarySource=install` in our self-hosted Renovate Docker app so that it still can upgrade old and new projects at the same time basically.

This worked great in general, but today I noticed that sometimes there are minor changes in `package-lock.json` when I run `npm install` locally again (with Node 16 / npm 8.3).
 
E.g. after a local npm install I get the following diff
```diff
--- a/package-lock.json
+++ b/package-lock.json
@@ -29,7 +29,7 @@
         "vue-loader": "^15.5.1",
         "vue-style-loader": "^4.1.2",
         "vue-template-compiler": "^2.5.22",
-        "webpack": "^5.70.0",
+        "webpack": "^5.0.0",
         "webpack-cli": "^4.0.0",
         "webpack-dev-server": "^4.7.3",
         "webpack-merge": "^5.0.0"
```

The logs revealed
```
DEBUG: Resolved version (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile)  DEBUG: Resolved version (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile
         "toolName": "npm",
         "constraint": undefined,
         "resolvedVersion": "8.5.3"
  DEBUG: Executing command (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile)  DEBUG: Executing command (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile
         "command": "install-tool npm 8.5.3"
  DEBUG: exec completed (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile)  DEBUG: exec completed (repository=<redacted>, branch=feature/renovate-crossborderflow-webpack-5.x-lockfile
         "cmd": "install-tool npm 8.5.3",
         "durationMs": 498,
         "stdout": "Installing legacy tool npm v8.5.3\n8.5.3\n",
         "stderr": ""
```

Apparently [Node 16 comes with npm 8.3](https://nodejs.org/en/download/) while [Node 17 comes with npm 8.5](https://nodejs.org/en/download/current/) and I _assume_ that this is the reason for my sporadical diffs.

So, finally, what I wanted to ask
- How does it detect 8.5.3 and wouldn't it better that it stays with the recommended LTS version and detect npm 8.3.x?
- I assume I can configure / prevent this via https://docs.renovatebot.com/node/#configuring-which-version-of-npm-renovate-uses or is there any better way?

Maybe this is also relevant for https://github.com/renovatebot/renovate/issues/14531 where `binarySource=install` will become the default.

**Comment by rarkins** [comment, marked as answer]
If there is no npm constraint (e.g. in `package.json>engines`, or in Renovate's own `constraints.npm` field), then Renovate picks the latest version of `npm` at the time of the run. If your project needs to constrain the version of npm used (e.g. `8.3.x` and no higher) then it needs to be in your engines or Renovate config.

**Reply by herndlm** [reply]
perfect thank you. I was not aware of `constraints.npm` for renovate which is the best solution for our use case then.

**Reply by herndlm** [reply]
I'm still not entirely sure if the diff is actually caused by the different npm version, but configuring it via constraints.npm definitely worked fine :)

**Reply by herndlm** [reply]
@rarkins sorry to keep bothering you, but do you have maybe any other ideas why those diffs are occurring? Renovate is now using npm 8.3.2, while I have 8.3.1 on my local system. In a PR where it updated postcss from 8.4.7 to 8.4.8 I get the following diff after my local `npm install`:
```diff
--- a/package-lock.json
+++ b/package-lock.json
@@ -26,7 +26,7 @@
         "debounce": "^1.2.0",
         "mini-css-extract-plugin": "^2.0.0",
         "node-sass": "^7.0.0",
-        "postcss": "^8.4.8",
+        "postcss": "^8.0.0",
         "sass-loader": "^12.0.0",
         "terser-webpack-plugin": "^5.0.0",
         "vue-loader": "^15.9.3",
``` 

It seems to be the always the same pattern. The package that was updated is modified in the lockfile that has the parts of the package.json. In the package.json I have `^8.0.0` which, I guess, should also end up in the lockfile there. But npm via renovate seems to be updating it to `^8.4.8`. Strange.

Just asking in case this is something extremely obvious I'm missing before I try to debug this more and/or create a reproducer :)

**Reply by rarkins** [reply]
It sounds like `rangeStrategy=update-lockfile` is not working as expected. I think you should be able to easily reproduce it, but I'm not sure there's an easy solution though.

**Reply by herndlm** [reply]
OK, I actually found out more about this. It seems to be unrelated to binarySource=install for sure. Maybe it is related to using a newer lockfile format which is why I didn't notice yet, not sure. It definitely also happened with Renovate 31.x.

Apparently renovate is modifying the package.json before running `npm install`:
```diff
--- a/package.json
+++ b/package.json
@@ -23,7 +23,7 @@
     "css-minimizer-webpack-plugin": "^3.0.0",
     "mini-css-extract-plugin": "^2.0.0",
     "node-sass": "^7.0.0",
-    "postcss": "^8.0.0",
+    "postcss": "^8.4.8",
     "sass-loader": "^12.0.0",
     "terser-webpack-plugin": "^5.0.0",
     "vue-loader": "^15.5.1",
```

I also see that before running `npm install` the lockfile is modified too. The `node_modules/postcss` block is removed.
Then it runs `npm install --package-lock-only --no-audit --ignore-scripts postcss@8.4.8` which results in the incorrect lockfile diff.
And then the lockfile is commited, but not package.json, which makes sense for my use case.

So, with my trivial understanding based on this, I would say it should not modify `package.json` when using `rangeStrategy=update-lockfile`. But maybe I'm overlooking something, I can debug later again. Would be interesting to know where it modifies it in case you're still reading this, but I'm not gonna explicitly tag you for now since I'm still play around :) Also my knowledge about npm is very limited, so I might be definitely missing something..

**Reply by rarkins** [reply]
If you can create a simple reproduction, I'd be interested in testing which series of edits/commands might work better. npm has indeed changed in this regard from v6->7->8

**Reply by herndlm** [reply]
OK this was easy to reproduce and looks like a simple `npm` limitation that might be hard to workaround.. I created https://github.com/renovatebot/renovate/issues/14584 for it to not mess up this discussion any further :)

**Comment by or-adar** [comment]
Hi @herndlm do you mind sharing whether your self-hosted Renovate instance is using `renovate/renovate:slim` image or `renovate/renovate`?
I'm asking this as I want to configure Renovate as well to allow it updating packages with different constraints, and I see in the docs that there are several approaches to this, so thought to ask you what was yours.

**Reply by herndlm** [reply]
We're using an image based (we just add some internal certificates) on `renovate/renovate`. And to allow different npm/composer versions of course, as mentioned in the first post, we're using `binarySource=install`.
